### DISCUSSION TITLE
(Host Side Question) Plugins from "Full Bucket Music" crash after "set_parent"

### DISCUSSION THREAD

**Discussion by Alpha21016** [free-audio/clap#386] [Q&A] [answered]
Hi, I'm developing a DAW software and I use QT6 as my GUI framework. I have implemented clap sdk and tried some plugins listed in this page "https://clapdb.tech/". Most of the plugins worked well, but all the plugins from "Full Bucket Music" crashed after I executed the function "bool(CLAP_ABI *set_parent)(const clap_plugin_t *plugin, const clap_window_t *window);". My codes to load the plugin follow the official project "clap-host", I don't know where I did it wrong. The following is my loading process:

1. load clap library
2. get clap_entry
3. get_factory
4. get_plugin_descriptor
5. create_plugin (use first desc->id)
6. init plugin
7. activate plugin
8. check can use gui and check if is in main thread
9. select os api = "win32" (win11/x64)
10. create gui using os_api
11. generate a clap_window_t
12. create an empty qt widget
13. pass the wid from this qt widget to clap_window_t (w.win32 = reinterpret_cast<clap_hwnd>(wid);)
14. execute set_parent
15. the set_parent function returns true and the code execution continues, after a few milliseconds my program crashes
16. exit code = -1073741819

I'd be very grateful If someone could tell me what is wrong with my process, thank you

**Comment by baconpaul** [comment]
Are you calling gui create before set parent?

https://github.com/free-audio/clap/blob/main/include/clap/ext/gui.h

that header contains the call order

**Reply by Alpha21016** [reply]
> Are you calling gui create before set parent?
> 
> https://github.com/free-audio/clap/blob/main/include/clap/ext/gui.h
> 
> that header contains the call order

Oh I forgot to write that in my question, thanks for your notice. Actually I did create gui between step 9 and 10, but it still crash. I compile and run the official demo project "clap-host", it works well with the "Full Bucket Music" plugins. I have compared each step with each other but I can hardly find the difference

**Reply by baconpaul** [reply]
Do other plugins like surge work in your host?

**Reply by Alpha21016** [reply]
Yes, I have tried SurgeXT,  PlexiTape, Odin2, Regency, Vital and all Fabfilter plugins, all of them successfully attach the qt widget and no crash happens.

**Comment by xris1658** [comment]
The exit code is 0xC0000005, which means an access violation, you should run your program with [Dr. Memory](https://drmemory.org/) to spot illegal memory accesses.

Since your program crashes on showing plugin GUI, and the load process itself seems correct, it's a good idea to check if your `clap_host` implementations, especially `clap_host_gui` callbacks are correct.

**Reply by baconpaul** [reply, marked as answer]
this is an excellent idea.

In fact, having your host print out / log all the host side extensions which are queried would help. It may be that FBM is querying an extension you don't support and is not being robust to null; or it may be that it queries one you partially support.

**Reply by Alpha21016** [reply]
Thanks for your help, I finally find out the reason. I did not fully implement the host callback function "CLAP_ABI *get_extension", when these plugins request extension such as "host_gui", I simply return nullptr and causes them crashed.

**Reply by baconpaul** [reply]
Right the clap contract is: if you support an extension you must support all methods of it. But a plugin should be resilient if you support none. Glad you found it!
